### PR TITLE
Feature #169 Admin theme support

### LINKED ISSUE #169: Portal theme
As a user I would like to have portal theme support

### PR DESCRIPTION
**Pull request by apachejuice**
Fixes #169, related to [metaform-api-spec #43](https://github.com/Metatavu/metaform-api-spec/issues/43)

### CODE REVIEW

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+     * @param data the data in the theme
+     * @param name the name of the theme
+     * @param slug the slug of the theme
+     * @param creatorId who created this theme
+     * @param lastModifierId who last modified this theme
+     *
+     * @return new [AdminTheme] object
+     */
+    fun create(
+        id: UUID, data: String, name: String, slug: String,
```
Comment (VottonDev) [inline src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt:30]: I'd probably put these into a row, so the code is more consistent with the style of the rest of the program. It's a preference but will be easier to pick up due to consistency.

Comment (VottonDev) [inline src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt:30]: I probably would put these into a row like the rest of the API, so it would follow a consistent code style making it easier for somebody else to pick up in the future.

**Review by VottonDev — COMMENTED** [review]
Nothing much to note as compiles fine and generally seems fine from the looks of it but would recommend to follow the same code style.

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        creatorId: UUID, lastModifierId: UUID
```
Comment (Chi-Chu319) [inline src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt:31]:
```suggestion
    fun create(
        id: UUID,
        data: String,
        name: String,
        slug: String,
        creatorId: UUID,
        lastModifierId: UUID
```

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    /**
+     * Finds an admin theme by id
+     *
+     * @param id the id of the admin theme
+     * @return the located admin theme or null
+     */
+    fun findById(id: UUID): AdminTheme? {
+        return adminThemeDAO.findById(id)
+    }
+
```
Comment (Chi-Chu319) [inline src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt:45]: doc

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+     * @param creatorId who created this theme
+     * @param lastModifierId who last modified this theme
+     *
+     * @return new [AdminTheme] object
+     */
+    fun create(
+        id: UUID, data: String, name: String, slug: String,
+        creatorId: UUID, lastModifierId: UUID
+    ): AdminTheme =
+        adminThemeDAO.create(id, data, name, slug, creatorId, lastModifierId)
```
Comment (Chi-Chu319) [inline src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt:33]: Always use curly braces to wrap function bodies

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/dao/AdminThemeDAO.kt [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        adminTheme.id = id
+        adminTheme.data = data
+        adminTheme.name = name
+        adminTheme.slug = slug
+        adminTheme.creatorId = creatorId
+        adminTheme.lastModifierId = lastModifierId
+
+        return persist(adminTheme)
+    }
+
```
Comment (Chi-Chu319) [inline src/main/kotlin/fi/metatavu/metaform/server/persistence/dao/AdminThemeDAO.kt:36]: Doc

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/model/AdminTheme.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+@Entity
+class AdminTheme : Metadata() {
+    @Id
+    var id: UUID? = null
+
+    @Column(nullable = false)
+    @NotEmpty
+    @NotNull
+    @Lob
+    lateinit var data: String
```
Comment (Chi-Chu319) [inline src/main/kotlin/fi/metatavu/metaform/server/persistence/model/AdminTheme.kt:23]: String serves the purpose very well

**Review by Chi-Chu319 — CHANGES_REQUESTED** [review]
Also the api impl and tests! always check the build before merging

File: src/main/kotlin/fi/metatavu/metaform/server/rest/AdminThemeApi.kt [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+            return createForbidden(createNotAllowedMessage(CREATE, ADMIN_THEME))
+        }
+
+        return createOk(adminThemeController.create(
+            UUID.randomUUID(),
+            adminTheme.data,
+            adminTheme.name,
+            adminTheme.slug,
+            userId,
+            userId,
```
Comment (VottonDev) [inline src/main/kotlin/fi/metatavu/metaform/server/rest/AdminThemeApi.kt:29]:
```suggestion
            userId
```
I think it's a duplicate?

**Review by VottonDev — CHANGES_REQUESTED** [review]
Noticed a duplicate value

**Comment by VottonDev** [comment]
The base tests have now been written I think but need to do the opposite to test exceptions as well.

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt (hunk 2 above)
Comment (VottonDev) [reply src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt]: Fixed.

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt (hunk 3 above)
Comment (VottonDev) [reply src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt:52]: Fixed.

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/dao/AdminThemeDAO.kt (hunk 5 above)
Comment (VottonDev) [reply src/main/kotlin/fi/metatavu/metaform/server/persistence/dao/AdminThemeDAO.kt:39]: Fixed. Added missing doc.

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+/**
+ * Controller for admin themes
+ */
+@ApplicationScoped
+class AdminThemeController {
+    @Inject
+    lateinit var adminThemeDAO: AdminThemeDAO
+
+    private val mapper = ObjectMapper()
```
Comment (Chi-Chu319) [inline src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt:18]:
```suggestion

    @Inject
    lateinit var mapper: ObjectMapper
```
using a injected mapper is better than creating one

File: src/main/kotlin/fi/metatavu/metaform/server/rest/AdminThemeApi.kt [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+     * 
+     * @return updated admin theme
+     */
+    override suspend fun updateAdminTheme(themeId: UUID, adminTheme: AdminTheme): Response {
+        loggedUserId ?: return createForbidden(UNAUTHORIZED)
+        if (!isRealmMetaformAdmin) {
+            return createForbidden(createNotAllowedMessage(CREATE, ADMIN_THEME))
+        }
+
+        val theme = adminThemeController.findById(themeId) ?: return createNotFound("No such admin theme: $themeId")
```
Comment (Chi-Chu319) [inline src/main/kotlin/fi/metatavu/metaform/server/rest/AdminThemeApi.kt:103]: use function createNotFoundMessage

File: src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+                    name = "Test admin theme",
+                    slug = "test-admin-theme"
+            )
+            
+            Assertions.assertNotNull(adminTheme)
+            Assertions.assertNotNull(adminTheme.id)
+            Assertions.assertNotNull(adminTheme.data)
+            Assertions.assertNotNull(adminTheme.name)
+            Assertions.assertNotNull(adminTheme.slug)
+        }
```
Comment (Chi-Chu319) [inline src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt:41]: Find the created theme and assert equal with adminTheme returned from the create test for all fields

Comment (Chi-Chu319) [reply src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt:41]: With that, you dont need a separate find test

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+     * Finds an admin theme by id
+     *
+     * @param id the id of the admin theme
+     * 
+     * @return the located admin theme or null
+     */
+    fun findById(id: UUID): AdminTheme? {
+        return adminThemeDAO.findById(id)
+    }
+
```
[the remark this answers is not in the document]
Comment (VottonDev) [reply src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt:61]: image

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt (hunk 8 above)
Comment (VottonDev) [reply src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt]: Changed.

File: src/main/kotlin/fi/metatavu/metaform/server/rest/AdminThemeApi.kt [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+     * 
+     * @return deleted admin theme
+     */
+    override suspend fun deleteAdminTheme(themeId: UUID): Response {
+        loggedUserId ?: return createForbidden(UNAUTHORIZED)
+        if (!isRealmMetaformAdmin) {
+            return createForbidden(createNotAllowedMessage(CREATE, ADMIN_THEME))
+        }
+
+        val theme = adminThemeController.findById(themeId) ?: return createNotFound("No such admin theme: $themeId")
```
[the remark this answers is not in the document]
Comment (VottonDev) [reply src/main/kotlin/fi/metatavu/metaform/server/rest/AdminThemeApi.kt]: Changed/Added.

File: src/main/kotlin/fi/metatavu/metaform/server/rest/AdminThemeApi.kt [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+     * 
+     * @return admin theme
+     */
+    override suspend fun findAdminTheme(themeId: UUID): Response {
+        loggedUserId ?: return createForbidden(UNAUTHORIZED)
+        if (!isRealmMetaformAdmin) {
+            return createForbidden(createNotAllowedMessage(CREATE, ADMIN_THEME))
+        }
+
+        val theme = adminThemeController.findById(themeId) ?: return createNotFound("No such admin theme: $themeId")
```
[the remark this answers is not in the document]
Comment (VottonDev) [reply src/main/kotlin/fi/metatavu/metaform/server/rest/AdminThemeApi.kt]: Changed/Added.

File: src/main/kotlin/fi/metatavu/metaform/server/rest/AdminThemeApi.kt (hunk 9 above)
Comment (VottonDev) [reply src/main/kotlin/fi/metatavu/metaform/server/rest/AdminThemeApi.kt]: Changed/Added.

File: src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+
+            val updatedAdminTheme = builder.metaformAdmin.adminThemes.update(adminTheme.id!!, secondAdminTheme)
+            val foundAdminTheme = builder.metaformAdmin.adminThemes.findById(adminTheme.id!!)
+            Assertions.assertNotNull(updatedAdminTheme)
+            Assertions.assertNotNull(updatedAdminTheme.id)
+            Assertions.assertNotNull(updatedAdminTheme.data)
+            Assertions.assertNotNull(updatedAdminTheme.name)
+            Assertions.assertNotNull(updatedAdminTheme.slug)
+            Assertions.assertEquals(updatedAdminTheme.data, foundAdminTheme.data)
+        }
```
Comment (Chi-Chu319) [inline src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt:68]: assert updated admin theme's properties equal with the secondAdminTheme's

File: src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+                    data = "data2",
+                    name = "Test admin theme2",
+                    slug = "test-admin-theme2"
+            )
+
+            Assertions.assertNotNull(adminTheme)
+            Assertions.assertNotNull(secondAdminTheme)
+            val adminThemes = builder.metaformAdmin.adminThemes.list()
+            Assertions.assertNotNull(adminThemes)
+            Assertions.assertEquals(4, adminThemes.size)
```
Comment (Chi-Chu319) [inline src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt:90]: Why is here 4 given that you created only 2 themes?

File: src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+    @Throws(Exception::class)
+    fun deleteAdminThemeTest() {
+        TestBuilder().use { builder ->
+            val adminTheme = builder.metaformAdmin.adminThemes.create(
+                    data = "data",
+                    name = "Test admin theme",
+                    slug = "test-admin-theme"
+            )
+
+            Assertions.assertNotNull(adminTheme)
```
Comment (Chi-Chu319) [inline src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt:104]: Try to find it and assert not null

File: src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+ * Tests for Admin Theme API
+ * @author Otto Hooper
+ */
+@QuarkusTest
+@QuarkusTestResource.List(
+        QuarkusTestResource(MysqlResource::class),
+        QuarkusTestResource(KeycloakResource::class)
+)
+@TestProfile(GeneralTestProfile::class)
+class AdminThemeTestsIT : AbstractTest() {
```
Comment (Chi-Chu319) [inline src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt:25]: Add unauthorized test for all the operations. e.g. create the admin theme with a non admin account and assert fail status with code 403

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+     * @param creatorId who created this theme
+     * @param lastModifierId who last modified this theme
+     *
+     * @return new [AdminTheme] object
+     */
+    fun create(
+        id: UUID,
+        data: Any?,
+        name: String,
+        slug: String,
```
Comment (Chi-Chu319) [inline src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt:37]: Change all the slug as nullable in controller, keep the slug as notnull in JPA.
In create admin theme, validate the slug with regex "^[a-z\\d]+(?:[-, _][a-z\\d]+)*$" and check for duplicates. if user provided slug is valid (not duplicated and matches regex), use the slug. if not valid, if not return e.g. invalidSlugException, catch it in api layer and return 400. if the slug is null, create the slug prefix from the name using slugify. if the slug prefix is not duplicated, use the prefix, else loop through all the numbers to check if e.g. slug-1 is duplicated, if not, use the slug

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+        theme: AdminTheme,
+        data: Any?,
+        name: String,
+        slug: String,
+    ): AdminTheme { 
+        return adminThemeDAO.update(
+            theme,
+            mapper.writeValueAsString(data),
+            name,
+            slug
```
Comment (Chi-Chu319) [inline src/main/kotlin/fi/metatavu/metaform/server/controllers/AdminThemeController.kt:73]: if user provided slug is the same as the found theme's slug or null, use the found slug. else check if the user provided is valid. if not return e.g. invalidSlugException, catch it in api layer and return 400

File: src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt (hunk 17 above)
Comment (Chi-Chu319) [inline src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt:25]: Add create with invalid slug test for duplicated slug and incorrectly formatted slug.

File: src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+            
+            val foundAdminTheme = builder.metaformAdmin.adminThemes.findById(adminTheme.id!!)
+            Assertions.assertNotNull(foundAdminTheme)
+            Assertions.assertNotNull(foundAdminTheme.id)
+            Assertions.assertNotNull(foundAdminTheme.data)
+            Assertions.assertNotNull(foundAdminTheme.name)
+            Assertions.assertNotNull(foundAdminTheme.slug)
+        }
+    }
+
```
Comment (Chi-Chu319) [inline src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt:44]: Add create theme with null slug test to see the slug in generated.

File: src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+            val updatedAdminTheme = builder.metaformAdmin.adminThemes.update(adminTheme.id!!, secondAdminTheme)
+            val foundAdminTheme = builder.metaformAdmin.adminThemes.findById(adminTheme.id!!)
+            Assertions.assertNotNull(updatedAdminTheme)
+            Assertions.assertNotNull(updatedAdminTheme.id)
+            Assertions.assertNotNull(updatedAdminTheme.data)
+            Assertions.assertNotNull(updatedAdminTheme.name)
+            Assertions.assertNotNull(updatedAdminTheme.slug)
+            Assertions.assertEquals(updatedAdminTheme.data, foundAdminTheme.data)
+        }
+    }
```
Comment (Chi-Chu319) [inline src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt:69]: Add update test with null slug and check is the slug remains the same

File: src/main/kotlin/fi/metatavu/metaform/server/rest/AdminThemeApi.kt [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+     * @param adminTheme admin theme to create
+     * 
+     * @return created admin theme
+     */
+    override suspend fun createAdminTheme(adminTheme: AdminTheme): Response {
+        val userId = loggedUserId ?: return createForbidden(UNAUTHORIZED)
+        if (!isRealmMetaformAdmin) {
+            return createForbidden(createNotAllowedMessage(CREATE, ADMIN_THEME))
+        }
+
```
Comment (Chi-Chu319) [inline src/main/kotlin/fi/metatavu/metaform/server/rest/AdminThemeApi.kt:29]: Add translator to translate all the admin theme JPA model to the rest model

File: src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt (hunk 14 above)
Comment (Chi-Chu319) [reply src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt:68]: With slug logic in place, this test should fail with duplicated slug. So, don't create the second theme, instead, just create a new rest model and update the theme with it. The slug should also be updated with a valid string.

File: src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+
```
Comment (Chi-Chu319) [inline src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt:70]: Add update test with invalid slug both duplicated and not in right format.

**Review by Chi-Chu319 — CHANGES_REQUESTED** [review]

File: src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt (hunk 15 above)
Comment (Chi-Chu319) [reply src/test/java/fi/metatavu/metaform/server/test/functional/tests/AdminThemeTestsIT.kt:91]: The test client should clean all the resources for each test. check the clean method in the test resource

File: src/test/java/fi/metatavu/metaform/server/test/functional/builder/impl/AdminThemeTestBuilderResource.kt [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+    /**
+     * Gets example theme data
+     *
+     * @return example theme data
+     */
+    val exampleThemeData: Map<String, String>
+        get() {
+            val versionData: MutableMap<String, String> = HashMap()
+            versionData["formData"] = "form value"
+            return versionData
```
Comment (Chi-Chu319) [inline src/test/java/fi/metatavu/metaform/server/test/functional/builder/impl/AdminThemeTestBuilderResource.kt:213]:
```suggestion
            val themeData: MutableMap<String, String> = HashMap()
            themeData["formData"] = "form value"
            return themeData
```
